doc/SAI-Proposal-VlanStack-QinQ.md

```diff
@@ -0,0 +1,380 @@
+
+Add support for Vlan Stack
+-------------------------------------------------------------------------------
+
+ Title       | Vlan Stack
+-------------|-----------------------------------------------------------------
+ Authors     | Edge-Core
+ Status      | In review
+ Type        | Standards track
+ Created     | 25/02/2022
+ Updated     | -
+ SAI-Version | 1.9.1
+
+-------------------------------------------------------------------------------
+
+## Overview ##
+
+The Vlan Stack feature includes two functions, vlan translation and QinQ.
+
+![qinq](figures/vlan_stack/vlan_stack_QinQ_usecase.png "Figure 1: QinQ usage scenario")
+__Figure 1: QinQ usage scenario__
+
+![vlan_translate](figures/vlan_stack/vlan_stack_vlan_xlate_usecase.png "Figure 2: Vlan translation usage scenario")
+__Figure 2: Vlan Translation usage scenario__
+
+## Packet Format ##
+
+Untagged packet, Single tagged packet and Double tagged packet
+
+![packet_format](figures/vlan_stack/TCPIP_802.1ad_DoubleTag.png "Figure 3: Packet packet format")
+__Figure 3: Packet format__
+
+## SAI Pipeline
+
+The packet processing pipeline is below:
+![vlan_stack-pipeline](figures/vlan_stack/vlan_stack_pipeline.png "Figure 4: Vlan Stack SAI Pipeline")
+__Figure 4: Vlan Stack SAI Pipeline__
+
+## SAI Attribute
+
+| SAI Component | SAI attribute                                  | Description                     |
+| ------------- | ---------------------------------------------- | ------------------------------- |
+| VLAN Stacking | SAI_VLAN_STACK_ATTR_STAGE                      | At which packet flow direction the action should take effect. Ingress or Egress    |
+|               | SAI_VLAN_STACK_ATTR_ACTION                     | Include SWAP, PUSH and POP operation    |
+|               | SAI_VLAN_STACK_ATTR_MATCH_TYPE                 | At which dot1q header position should be used to match original vlan ID. Inner or Outer. Please note that for a PUSH action, this should always set to Inner.    |
```

Comment:
Could you please enhance this to support matching multi-tag ?
Example : `Match (OVID-100/IVID-10) translate to SVID-200 and vice versa`

Reply:
I will enhance this in the next commit ASAP.

Reply:
I remove the attribute, SAI_VLAN_STACK_ATTR_MATCH_TYPE. And add a new data type, sai_vlan_stacking_vid_t, to match header position and vid in the same time.
For your example, please refer "Ingress/Swap on double tag from single tag" and "Egress/Swap on single tag from double tag" to achieve the goal.